sway-core/src/type_system/info.rs

```diff
@@ -412,6 +412,65 @@ impl OrdWithEngines for TypeInfo {
 }
 
 impl DisplayWithEngines for TypeInfo {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>, engines: Engines<'_>) -> fmt::Result {
+        use TypeInfo::*;
+        let s = match self {
+            Unknown => "unknown".into(),
+            UnknownGeneric { name, .. } => name.to_string(),
+            Placeholder(type_param) => type_param.name_ident.to_string(),
+            TypeParam(n) => format!("{n}"),
+            Str(x) => format!("str[{}]", x.val()),
+            UnsignedInteger(x) => match x {
+                IntegerBits::Eight => "u8",
+                IntegerBits::Sixteen => "u16",
+                IntegerBits::ThirtyTwo => "u32",
+                IntegerBits::SixtyFour => "u64",
+            }
+            .into(),
+            Boolean => "bool".into(),
+            Custom { call_path, .. } => call_path.suffix.to_string(),
+            Tuple(fields) => {
+                let field_strs = fields
+                    .iter()
+                    .map(|field| format!("{:?}", engines.help_out(field)))
+                    .collect::<Vec<String>>();
+                format!("({})", field_strs.join(", "))
+            }
+            SelfType => "Self".into(),
+            B256 => "b256".into(),
+            Numeric => "numeric".into(),
+            Contract => "contract".into(),
+            ErrorRecovery => "unknown".into(),
+            Enum(decl_ref) => {
+                let decl = engines.de().get_enum(decl_ref);
+                print_inner_types(
+                    engines,
+                    decl.call_path.suffix.as_str().to_string(),
+                    decl.type_parameters.iter().map(|x| x.type_id),
+                )
+            }
+            Struct(decl_ref) => {
+                let decl = engines.de().get_struct(decl_ref);
+                print_inner_types(
+                    engines,
+                    decl.call_path.suffix.as_str().to_string(),
+                    decl.type_parameters.iter().map(|x| x.type_id),
+                )
+            }
+            ContractCaller { abi_name, .. } => format!("ContractCaller<{abi_name}>"),
+            Array(elem_ty, count) => {
+                format!("[{:?}; {}]", engines.help_out(elem_ty), count.val())
+            }
+            Storage { .. } => "storage".into(),
+            RawUntypedPtr => "pointer".into(),
+            RawUntypedSlice => "slice".into(),
```

Comment:
I'm not sure what the best user-friendly label for these is. Any example code using these types would be helpful!

Reply:
Yeah I think this fine!